### PR TITLE
feat(api): add preferences entity and schema

### PR DESCRIPTION
**Pull request by davidsoderberg**

### CODE REVIEW

File: libs/application-generic/src/usecases/write-preferences/write-preferences.usecase.ts
Comment (rifont) [inline libs/application-generic/src/usecases/write-preferences/write-preferences.usecase.ts:1]: we can rename this to `upsert-preferences` to provide a semantic description of the use-case

File: libs/application-generic/src/usecases/get-preferences/get-preferences.command.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { EnvironmentCommand } from '../../commands';
+
+export class GetPreferencesCommand extends EnvironmentCommand {
+  readonly subscriberId: string;
```
Comment (rifont) [inline libs/application-generic/src/usecases/get-preferences/get-preferences.command.ts]: I think this should be typed as an optional param.

File: apps/api/src/bootstrap.ts [hunk 2]
```diff
@@ -89,7 +89,13 @@ export async function bootstrap(expressApp?): Promise<INestApplication> {
   app.use(helmet());
   app.enableCors(corsOptionsDelegate);
 
-  app.setGlobalPrefix(CONTEXT_PATH + 'v1');
+  app.setGlobalPrefix(CONTEXT_PATH);
```
Comment (davidsoderberg) [inline apps/api/src/bootstrap.ts]: This and the changes below allow us to in the future add new versions to our api

**Comment by davidsoderberg** [comment]
Needs to add usage of preferences in sync usecase

File: apps/api/src/app/inbox/inbox.module.ts [hunk 3]
```diff
@@ -1,15 +1,18 @@
...
+import { PreferencesRepository } from '@novu/dal';
 import { AuthModule } from '../auth/auth.module';
-import { SubscribersModule } from '../subscribers/subscribers.module';
 import { IntegrationModule } from '../integrations/integrations.module';
+import { SharedModule } from '../shared/shared.module';
+import { SubscribersModule } from '../subscribers/subscribers.module';
+import { InboxController } from './inbox.controller';
+import { USE_CASES } from './usecases';
+
+const EXTERNAL_USE_CASES = [GetPreferences, UpsertPreferences, PreferencesRepository];
```
Comment (rifont) [inline apps/api/src/app/inbox/inbox.module.ts]: I think we should specify `PreferencesModule` in the `imports` array instead to fully encapsulate the use-cases. This is the Nest.js recommended approach and ensures we don't need to come back to this file if we update the list of use-cases.

File: apps/api/src/app/preferences/dtos/preferences.dto.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import { Type } from 'class-transformer';
+import { IsBoolean, ValidateNested } from 'class-validator';
+
+class Preference {
```
Comment (rifont) [inline apps/api/src/app/preferences/dtos/preferences.dto.ts]:
```suggestion
export class Preference {
```
We need to export all nested DTOs to ensure that `@nestjs/swagger` can access all the nested typings when it produces the OpenAPI spec.

File: apps/api/src/app/preferences/dtos/preferences.dto.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  @IsBoolean()
+  defaultValue: boolean;
+
+  @IsBoolean()
+  readOnly: boolean;
+}
+
+class Channels {
```
Comment (rifont) [inline apps/api/src/app/preferences/dtos/preferences.dto.ts]: As above

```suggestion
export class Channels {
```

File: apps/api/src/app/preferences/preferences.controller.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        organizationId: user.organizationId,
+        key: FeatureFlagsKeysEnum.IS_WORKFLOW_PREFERENCES_ENABLED,
+      })
+    );
+
+    if (!isEnabled) {
+      throw new NotFoundException();
+    }
+  }
+}
```
Comment (rifont) [inline apps/api/src/app/preferences/preferences.controller.ts:80]: 👏 Great use of method encapsulation

File: apps/api/src/app/preferences/preferences.spec.ts
Comment (rifont) [inline apps/api/src/app/preferences/preferences.spec.ts:1]: Very thorough tests, well done 👏

File: apps/api/src/app/subscribers/subscribers.module.ts [hunk 7]
```diff
@@ -1,15 +1,19 @@
...
-import { SubscribersController } from './subscribers.controller';
+import { GetPreferences, UpsertPreferences } from '@novu/application-generic';
+import { PreferencesRepository } from '@novu/dal';
 import { AuthModule } from '../auth/auth.module';
+import { SharedModule } from '../shared/shared.module';
 import { WidgetsModule } from '../widgets/widgets.module';
+import { SubscribersController } from './subscribers.controller';
+import { USE_CASES } from './usecases';
+
+const EXTERNAL_USE_CASES = [GetPreferences, UpsertPreferences, PreferencesRepository];
```
Comment (rifont) [inline apps/api/src/app/subscribers/subscribers.module.ts]: Same goes here, we should add `PreferencesModule` to the `imports` array to encapsulate all the use-cases

File: apps/api/src/app/workflows/workflow.module.ts [hunk 8]
```diff
@@ -1,17 +1,21 @@
...
+import { USE_CASES } from './usecases';
 import { WorkflowController } from './workflow.controller';
 
+const EXTERNAL_USE_CASES = [GetPreferences, UpsertPreferences, PreferencesRepository];
+
 @Module({
   imports: [SharedModule, MessageTemplateModule, ChangeModule, AuthModule, IntegrationModule],
   controllers: [NotificationTemplateController, WorkflowController],
-  providers: [...USE_CASES],
+  providers: [...USE_CASES, ...EXTERNAL_USE_CASES],
```
Comment (rifont) [inline apps/api/src/app/workflows/workflow.module.ts]: Same here, replace by adding `PreferencesModule` to `imports` array

File: apps/api/src/bootstrap.ts [hunk 9]
```diff
@@ -3,7 +3,7 @@ import 'newrelic';
 import '@sentry/tracing';
 
 import helmet from 'helmet';
-import { INestApplication, Logger, ValidationPipe } from '@nestjs/common';
+import { INestApplication, Logger, ValidationPipe, VersioningType } from '@nestjs/common';
```
Comment (rifont) [inline apps/api/src/bootstrap.ts]: ☑ **chore:** Remove unused import‏
```suggestion
import { INestApplication, Logger, ValidationPipe } from '@nestjs/common';
```

File: apps/worker/src/app/workflow/workflow.module.ts [hunk 10]
```diff
@@ -135,6 +136,7 @@ const USE_CASES = [
   InboundEmailParse,
   ExecutionLogRoute,
   ExecuteBridgeJob,
+  GetPreferences,
```
Comment (rifont) [inline apps/worker/src/app/workflow/workflow.module.ts:139]: We can import the `PreferencesModule` here too.

File: libs/application-generic/src/usecases/get-preferences/get-preferences.usecase.ts [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    command: GetPreferencesCommand,
+  ): Promise<WorkflowOptionsPreferences> {
+    const isEnabled = await this.getFeatureFlag.execute(
+      GetFeatureFlagCommand.create({
+        userId: 'system',
+        environmentId: command.environmentId,
+        organizationId: command.organizationId,
+        key: FeatureFlagsKeysEnum.IS_WORKFLOW_PREFERENCES_ENABLED,
+      }),
+    );
```
Comment (rifont) [inline libs/application-generic/src/usecases/get-preferences/get-preferences.usecase.ts:31]: We had a feature flag check on the controller too. Given that we have this check on the use-case (which would be my preference, to co-locate the FF with the feature), I think we can remove the controller check, right?

File: libs/application-generic/src/usecases/get-preferences/get-preferences.usecase.ts [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        in_app:
+          result.channels.in_app.defaultValue || result.workflow.defaultValue,
+        sms: result.channels.sms.defaultValue || result.workflow.defaultValue,
+        email:
+          result.channels.email.defaultValue || result.workflow.defaultValue,
+        push: result.channels.push.defaultValue || result.workflow.defaultValue,
+        chat: result.channels.chat.defaultValue || result.workflow.defaultValue,
+      };
+    } catch (e) {
+      return undefined;
```
Comment (rifont) [inline libs/application-generic/src/usecases/get-preferences/get-preferences.usecase.ts]: ☑ **chore:** I think we should throw an error here instead. Why are we returning undefined if an error presents? If we do need it, let's add a code comment to highlight why.‏

File: libs/dal/src/repositories/preferences/preferences.entity.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+type WorkflowOptionsPreferences = {
+  workflow: PreferenceOptions;
+  channels: Record<ChannelTypeEnum, PreferenceOptions>;
+};
+
+export enum PreferencesTypeEnum {
+  SUBSCRIBER_GLOBAL = 'SUBSCRIBER_GLOBAL',
+  SUBSCRIBER_WORKFLOW = 'SUBSCRIBER_WORKFLOW',
+  USER_WORKFLOW = 'USER_WORKFLOW',
+  CODE_FIRST_WORKFLOW = 'CODE_FIRST_WORKFLOW',
```
Comment (rifont) [inline libs/dal/src/repositories/preferences/preferences.entity.ts]: The naming is semantic, but I think we can do better to abstract away from the V2 experience. How about the following?

```suggestion
  WORKFLOW_RESOURCE = 'WORKFLOW_RESOURCE',
```

**Review by rifont — APPROVED** [review]
Amazing work David! 🚀

File: libs/application-generic/src/usecases/get-preferences/get-preferences.usecase.ts [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+      (item) => item.type === PreferencesTypeEnum.USER_WORKFLOW,
+    );
+  }
+
+  private getWorkflowPreferences(items: PreferencesEntity[]) {
+    return items.find(
+      (item) => item.type === PreferencesTypeEnum.CODE_FIRST_WORKFLOW,
+    );
+  }
+
```
Comment (antonjoel82) [inline libs/application-generic/src/usecases/get-preferences/get-preferences.usecase.ts]: 💭 **thought (non-blocking):** Now that we have these enum values, we could just make a `getPreferenceByType` function that takes the array and the type.

Also, these methods could either be declared as external helper functions, or static methods‏

File: apps/api/src/app/preferences/dtos/preferences.dto.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  @Type(() => Preference)
+  sms: Preference;
+
+  @ValidateNested({ each: true })
+  @Type(() => Preference)
+  chat: Preference;
+
+  @ValidateNested({ each: true })
+  @Type(() => Preference)
+  push: Preference;
```
Comment (antonjoel82) [inline apps/api/src/app/preferences/dtos/preferences.dto.ts]: ❓ **question:** Can these also be keyed by the enum?‏

File: apps/api/src/app/inbox/usecases/update-preferences/update-preferences.usecase.ts [hunk 16]
```diff
@@ -171,4 +195,63 @@ export class UpdatePreferences {
...
+        push: {
+          defaultValue: item.channels.push || PREFERENCE_DEFAULT_VALUE,
+          readOnly: false,
+        },
+        chat: {
+          defaultValue: item.channels.chat || PREFERENCE_DEFAULT_VALUE,
+          readOnly: false,
+        },
+      },
+    };
```
Comment (antonjoel82) [inline apps/api/src/app/inbox/usecases/update-preferences/update-preferences.usecase.ts:234]: 👏 **praise:** Thank for the changes here!‏

**Review by antonjoel82 — APPROVED** [review]
Awesome! Good job on the updates -- I think this makes things easier to use and maintains encapsulation well

File: apps/worker/src/app/workflow/workflow.module.ts (hunk 10 above)
Comment (davidsoderberg) [reply apps/worker/src/app/workflow/workflow.module.ts:139]: Should we do it here as it is not in the api service?

File: libs/application-generic/src/usecases/get-preferences/get-preferences.usecase.ts (hunk 11 above)
Comment (davidsoderberg) [reply libs/application-generic/src/usecases/get-preferences/get-preferences.usecase.ts:31]: No as we do not do this in upsert, but maybe we should do it there to?

File: apps/worker/src/app/workflow/workflow.module.ts (hunk 10 above)
Comment (rifont) [reply apps/worker/src/app/workflow/workflow.module.ts:139]: Ah fair point, let's keep Worker as is for now. But in the long term, we should have a way to share such modules across both API and Worker.

File: libs/application-generic/src/usecases/get-preferences/get-preferences.usecase.ts (hunk 11 above)
Comment (rifont) [reply libs/application-generic/src/usecases/get-preferences/get-preferences.usecase.ts:31]: Yes, I think it makes sense to remain consistent.

**Pull request merged**